src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs

```diff
@@ -362,7 +368,15 @@ public override object Invoke(ServiceProvider provider)
 
             public override Expression Build(Expression provider)
             {
-                return base.Build(Expression.Field(provider, "_root"));
+                // PERF: Singleton services once created will always be available in the lookup dictionary
+                // Hence it is cheaper to directly call its Invoke method that will do the lookup rather than 
+                // building an expression tree that does the same.
+                // This change reduces the memory allocations during the start up because, expressions accessing private members
+                // involves expensive code access security checks for Desktop CLR.
+                return Expression.Call(
+                    Expression.Constant(this, typeof(SingletonCallSite)),
+                    InvokeMethodInfo,
+                    provider);
```

Comment:
This actually looks pretty safe and reasonable. Calling `Invoke()` instead of doing the equivalent action directly in the expression tree isn't really that costly since we'll always be pulling the instance out of a dictionary anyway instead of doing any complicated reflection, constructor selection, etc... If calling `Invoke()` avoids costly CAS-related allocations, that seems better to me than some other sugested alternatives such as making `_root` public.

@davidfowl The new compiled expression returned by this method, `SingletonCallSite.Build`, calls into `ScopedCallSite.Invoke` which would add a newly created singleton to `_resolvedServices` (used in part to track disposables) if it wasn't there already. Even though it's not really possible as @pranavkm mentions for the singleton not to be in the dictionary yet, the code would still be correct (albeit slow) if that wasn't the case.